db/python/layers/cohort.py

```diff
@@ -0,0 +1,184 @@
+            cohort_name: str,
+            dry_run: bool,
+            cohort_criteria: CohortCriteria = None,
+            template_id: int = None,
+    ):
+        """
+        Create a new cohort from the given parameters. Returns the newly created cohort_id.
+        """
+
+        create_cohort_template = True
+
+        # Input validation
+        if not cohort_criteria and not template_id:
+            raise ValueError('A cohort must have either criteria or be derived from a template')
+
+        # Get template from ID
+        template: dict[str, str] = {}
+        if template_id:
+            template = await self.ct.get_cohort_template(template_id)
+            if not template:
+                raise ValueError(f'Cohort template with ID {template_id} not found')
+
+        if template and cohort_criteria:
+            # TODO: Handle this case. For now, not supported.
+            raise ValueError('A cohort cannot have both criteria and be derived from a template')
+
+        # Only provide a template id
+        if template and not cohort_criteria:
+            create_cohort_template = False
+            criteria_dict = json.loads(template['criteria'])
+            cohort_criteria = CohortCriteria(**criteria_dict)
+
+        projects_to_pull = await self.pt.get_and_check_access_to_projects_for_names(
+            user=self.connection.author, project_names=cohort_criteria.projects, readonly=True
+        )
+        projects_to_pull = [p.id for p in projects_to_pull]
+
+        sg_filter = get_sg_filter(
+            projects=projects_to_pull,
+            sg_ids_internal_rich=cohort_criteria.sg_ids_internal,
+            excluded_sgs_internal_rich=cohort_criteria.excluded_sgs_internal,
+            sg_technology=cohort_criteria.sg_technology,
+            sg_platform=cohort_criteria.sg_platform,
+            sg_type=cohort_criteria.sg_type
+        )
+
+        sgs = await self.sglayer.query(sg_filter)
+
+        rich_ids = sequencing_group_id_format_list([sg.id for sg in sgs])
+
+        if dry_run:
+            return {'dry_run': True, 'template_id': template_id or 'CREATE NEW', 'sequencing_group_ids': rich_ids}
+
+        # 2. Create cohort template, if required.
+        if create_cohort_template:
+            cohort_template = CohortTemplate(
+                name=cohort_name,
+                description=description,
+                criteria=cohort_criteria
+            )
+            template_id = await self.create_cohort_template(
+                cohort_template=cohort_template,
+                project=project_to_write
+            )
+
+        assert template_id, 'Template ID must be set'
+
+        # 3. Create Cohort
+        cohort_id = await self.ct.create_cohort(
+            project=project_to_write,
+            cohort_name=cohort_name,
+            sequencing_group_ids=[sg.id for sg in sgs],
+            description=description,
+            author=author,
+            template_id=template_id,
```

Comment:
Seems that `template_id` refers to the CohortTemplate entry, i.e., a row in the `cohort_template` table.

Which would mean we need to update the db constraint: `FK_COHORT_DERIVED_FROM_ID_COHORT` has this field referencing `cohort(id)` but it looks like it should be `cohort_template(id)`.

Maybe we can just delete the constraint and add a new one called FK_COHORT_TEMPLATE_ID_ID_COHORT_TEMPLATE or something like that.

Reply:
Yes I've been trying to fix this since yesterday, but have found myself in a very bad liquibase state so it's taken me longer to push the change out, sorry! Trying to roll back some bad the changes now so I can apply the fix.